types/license.go

```diff
@@ -0,0 +1,165 @@
+package types
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/kong/deck/crud"
+	"github.com/kong/deck/state"
+	"github.com/kong/go-kong/kong"
+)
+
+// licenseCRUD implements crud.Actions interface.
+type licenseCRUD struct {
+	client *kong.Client
+}
+
+func licenseFromStruct(arg crud.Event) *state.License {
+	license, ok := arg.Obj.(*state.License)
+	if !ok {
+		panic("unexpected type, expected *state.License")
+	}
+	return license
+}
+
+// Create creates a License in Kong.
+// The arg should be of type crud.Event, containing the license to be created,
+// else the function will panic.
+// It returns a the created *state.License.
+func (s *licenseCRUD) Create(ctx context.Context, arg ...crud.Arg) (crud.Arg, error) {
+	event := crud.EventFromArg(arg[0])
+	license := licenseFromStruct(event)
+	createdLicense, err := s.client.Licenses.Create(ctx, &license.License)
+	if err != nil {
+		return nil, err
+	}
+	return &state.License{License: *createdLicense}, nil
+}
+
+// Delete deletes a License in Kong.
+// The arg should be of type crud.Event, containing the license to be deleted,
+// else the function will panic.
+// It returns a the deleted *state.License.
+func (s *licenseCRUD) Delete(ctx context.Context, arg ...crud.Arg) (crud.Arg, error) {
+	event := crud.EventFromArg(arg[0])
+	license := licenseFromStruct(event)
+	err := s.client.Licenses.Delete(ctx, license.ID)
+	if err != nil {
+		return nil, err
+	}
+	return license, nil
+}
+
+// Update updates a License in Kong.
+// The arg should be of type crud.Event, containing the license to be updated,
+// else the function will panic.
+// It returns a the updated *state.License.
+func (s *licenseCRUD) Update(ctx context.Context, arg ...crud.Arg) (crud.Arg, error) {
+	event := crud.EventFromArg(arg[0])
```

Comment:
- Can we check if `arg` has at least 1 element?
- Why not just use `crud.Event` type?

Reply:
This is yet more copy pasta where every entity type in deck looks like this, but in this case I can't offer much any explanation as to why it's like this. The boolean positional arguments are at least a simple design that probably grew naturally out of initially only having one toggle.

This I can see that `crud.Arg` is an `interface{}` alias, which it probably doesn't need to be, and that I don't immediately see from [relevant code in the differ](https://github.com/Kong/deck/blob/a7cf38f623ad488cfffc1ecab21f32043fe7c301/diff/diff.go#L302-L352) where it'd ever be anything other than an Event or when you'd have more than one Event arg. I don't see anything else being passed as args.

The arg length check isn't going to get us much other than a cleaner error that's not user actionable, as any case where it's _not_ at least one item points to a bug further up. Your only recourse is to deliver replication steps to us and have us figure out what we did wrong in the diff library. Panicking on out of bounds is correctly indicating that there's something horribly wrong elsewhere in our code.

My best guess is that this was intended for some sort of expanded functionality beyond Events, but that said functionality was unknown at the time of design, and ultimately it turned out to be a YAGNI expansion port.

No follow-up issue here since I can't even point to a clearly good refactor, and would probably suggest a broader redesign of the differ. The type-related code is at least simple if ugly, whereas the differ code that calls it is confusing, difficult to debug, and apparently full of unnecessary room for future expansion that does nothing but make it impossible to have stronger compile-time validation that would suffice for what deck actually needs to do.